sourcepolicy/engine.go

```diff
@@ -0,0 +1,226 @@
+	ErrSourceDenied = errors.New("source denied by policy")
+
+	// ErrTooManyOps is returned by the policy engine when there are too many converts for a single source op.
+	ErrTooManyOps = errors.New("too many operations")
+)
+
+// Engine is the source policy engine.
+// It is responsible for evaluating a source policy against a source operation.
+// Create one with `NewEngine`
+//
+// Rule matching is delegated to the `Matcher` interface.
+// Mutations are delegated to the `Mutater` interface.
+type Engine struct {
+	pol     []*spb.Policy
+	matcher Matcher
+	mutater Mutater
+	sources map[string]*Source
+}
+
+// NewEngine creates a new source policy engine.
+func NewEngine(pol []*spb.Policy, matcher Matcher, mutater Mutater) *Engine {
+	if matcher == nil {
+		matcher = DefaultMatcher
+	}
+	if mutater == nil {
+		mutater = DefaultMutater
+	}
+	return &Engine{
+		pol:     pol,
+		matcher: matcher,
+		mutater: mutater,
+	}
+}
+
+// TODO: The key here can't be used to cache attr constraint regexes.
+func (e *Engine) source(src *spb.Source) *Source {
+	if e.sources == nil {
+		e.sources = map[string]*Source{}
+	}
+
+	key := src.MatchType.String() + " " + src.Type + "://" + src.Identifier
+
+	if s, ok := e.sources[key]; ok {
+		return s
+	}
+	s := &Source{Source: src}
+
+	e.sources[key] = s
+	return s
+}
+
+// Evaluate evaluates a source operation against the policy.
+//
+// Policies are re-evaluated for each convert rule.
+// Evaluate will error if the there are too many converts for a single source op to prevent infinite loops.
+// This function may error out even if the op was mutated, in which case `true` will be returned along with the error.
+//
+// An error is returned when the source is denied by the policy.
+func (e *Engine) Evaluate(ctx context.Context, op *pb.Op) (bool, error) {
+	if len(e.pol) == 0 {
+		return false, nil
+	}
+
+	var (
+		st      evalState
+		mutated bool
+	)
+	const maxIterr = 20
+
+	for i := 0; ; i++ {
+		if i > maxIterr {
+			return mutated, errors.Wrapf(ErrTooManyOps, "too many mutations on a single source")
+		}
+
+		srcOp := op.GetSource()
+		if srcOp == nil {
+			return false, nil
+		}
+		if i == 0 {
+			ctx = bklog.WithLogger(ctx, bklog.G(ctx).WithField("orig", *srcOp).WithField("updated", op.GetSource()))
+		}
+
+		mut, err := e.evaluatePolicies(ctx, srcOp, &st)
+		if mut {
+			mutated = true
+		}
+		if err != nil {
+			return mutated, err
+		}
+		if !mut {
+			break
+		}
+	}
+
+	return mutated, nil
+}
+
+func (e *Engine) evaluatePolicies(ctx context.Context, srcOp *pb.SourceOp, st *evalState) (bool, error) {
+	ident := srcOp.GetIdentifier()
+	scheme, ref, found := strings.Cut(ident, "://")
+	if !found || ref == "" {
+		return false, errors.Errorf("failed to parse %q", ident)
+	}
+
+	ctx = bklog.WithLogger(ctx, bklog.G(ctx).WithFields(map[string]interface{}{
+		"scheme": scheme,
+		"ref":    ref,
+	}))
+
+	for _, pol := range e.pol {
+		mut, err := e.evaluatePolicy(ctx, pol, srcOp, st, scheme, ref)
```

Comment:
I think we should try and evaluate each policy independently from the others, though I'm not sure on this.

I guess it depends on how we want different policies to interact together -  as written, a policy can be satisfied, even if another policy would produce a conflicting result. e.g. my frontend policy might say I allow library/alpine, while my buildkit backend might say I only allow myorg/* (confusingly, this would fail depending on which way round it's specified). In this case, if multiple policies produce conflicting results, we should exit, and refuse to build - otherwise, we're implicitly ranking different sources of policy, and requiring the user to understand which ones have precedence.

Multiple `CONVERT` rules are harder, but I think the choice to just pick the first conversion seems alright.

@tonistiigi WDYT?

Reply:
I think that's pretty much how this works because `evaluatePolicy` returns an error when a source is denied and that bubbles up.
The only change is to not share evalState between policies (and potentially we don't even need evalState depending on other questions about how we want to handle allow->deny/deny->allow)

Reply:
Cool, this is updated. Note that, at least for the built-in engine, this makes `ALLOW` effectively a no-op.